core/modules/replica/ConfigurationExceptions.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_REPLICA_CONFIGURATIONEXCEPTIONS_H
+#define LSST_QSERV_REPLICA_CONFIGURATIONEXCEPTIONS_H
+
+/**
+ * This header defines classes thrown as exceptions on Configuration-specific
+ * failures.
+ */
+
+// System headers
+#include <stdexcept>
+#include <string>
+
+// This header declarations
+namespace lsst {
+namespace qserv {
+namespace replica {
+/**
+ * The class ConfigError is the base class representing exceptions thrown by
+ * the Configuration service.
+ */
+class ConfigError: public std::runtime_error {
+public:
+    using std::runtime_error::runtime_error;
+};
+
+/**
+ * The class ConfigVersionMismatch represents exceptions thrown on the expected versus
+ * actual version mismatch of the configuration found in the peristent store.
+ */
+class ConfigVersionMismatch: public ConfigError {
+public:
+    int const version;
+    int const requiredVersion;
+    explicit ConfigVersionMismatch(int version_, int requiredVersion_)
+        :   ConfigError("Configuration version " + std::to_string(version_) + " found in"
+                        + " the persistent state or a JSON object doesn't match the required"
+                        + " version " + std::to_string(requiredVersion_) + "."),
+            version(version_),
+            requiredVersion(requiredVersion_) {
+    }
+};
+
+/**
+ * The class ConfigTypeMismatch represents exceptions thrown during type conversions
+ * of the parameter values if the expected type of a parameter doesn't match the actual
+ * one stored in the configuration.
+ */
+class ConfigTypeMismatch: public ConfigError {
+public:
+    using ConfigError::ConfigError;
+};
```

Comment:
The only location where this exception is thrown is:
```
// file: Configuration.h
    template <typename T>
    T get(std::string const& category, std::string const& param) const {
        util::Lock const lock(_mtx, _context(__func__));
        try {
            return _get(lock, category, param).get<T>();
        } catch (std::invalid_argument const&) {
            throw;
        } catch (std::exception const& ex) {
            throw ConfigTypeMismatch(
                _context(__func__) + " failed to convert the parameter for category '"
                + category + "', and param '" + param + "', ex: " + std::string(ex.what()) + "'.");
        }
    }
```
It's possible to get the type name from a JSON object returned by the method `_get(lock, category, param)` using `json::type()` and storying it within the exception. However, I'm not sure if anyone would get anything useful out of C++11:
```
#include <typeinfo>
int a;
std::cout << typeid(a).name() << '\n';
```
I remember it would print some gibberish, or perhaps some single letter for primitive types (like `i` for integers, etc.). Though, knowing a type required by the schema for the attempted parameter may be useful. I will simply extend the message (see the code above) that's used for constructing the exception to carry the expected type.